:memo: Prepared Documentation for the 2.0 Realease (#597)
munich-quantum-toolkit/bench: docs/quickstart.md

    level=BenchmarkLevel.MAPPED,
    circuit_size=5,
    target=get_device("ionq_forte_36"),
    opt_level=2,
)
qc_mapped_level.draw(output="mpl")
```

## Mirror Circuits

```{code-cell} ipython3

> qc_mirrored = get_benchmark(

Reviewer: Could we draw both the original circuit additionally to the mirrored one?

Author: Makes sense.

    benchmark="ghz",
    level=BenchmarkLevel.MAPPED,
    circuit_size=3,
    target=get_device("ibm_falcon_27"),
    opt_level=0,
    # generate_mirror_circuit=True,
)
qc_mirrored.draw(output="mpl")
```

> ## Self-defined Circuits

Reviewer: A short description would help here as well.

```{code-cell} ipython3
from qiskit import QuantumCircuit

qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)

> qc_circuit = get_benchmark(benchmark=qc, level=BenchmarkLevel.NATIVEGATES, target=get_device("ionq_forte_36"))

Reviewer: Although this works, I would prefer to use a native gateset target instead of a device for the native gates level.

qc_circuit.draw(output="mpl")
```

> ## Self-defined Targets

Reviewer: Same.

```{code-cell} ipython3
from qiskit.providers.fake_provider import GenericBackendV2

standard_gates = ["id", "x", "sx", "rz", "cx"]
backend = GenericBackendV2(num_qubits=5, basis_gates=standard_gates)
target = backend.target
target.description = "Awesome Target"

qc_target = get_benchmark(
    benchmark="dj",
    level=BenchmarkLevel.NATIVEGATES,
    circuit_size=5,
    target=target,
    opt_level=2,
)
qc_target.draw(output="mpl")
```

> ## Random and Symbolic Parameters

Reviewer: A short description would be good.

### Random Parameters (Default)

```{code-cell} ipython3
qc_random = get_benchmark(benchmark="qaoa", level=BenchmarkLevel.ALG, circuit_size=2)
qc_random.draw(output="mpl")
```

### Symbolic Parameters

```{code-cell} ipython3
qc_symbolic = get_benchmark(benchmark="qaoa", level=BenchmarkLevel.ALG, circuit_size=2, random_parameters=False)
qc_symbolic.draw(output="mpl")
```

> ## Output Formats

Reviewer: Same here.

    level=BenchmarkLevel.MAPPED,
    circuit_size=5,
    target=get_device("ionq_forte_36"),
    opt_level=2,
)
qc_mapped_level.draw(output="mpl")
```

## Mirror Circuits

Mirror circuits replicate a given benchmark circuit and its mirror image (inverse) concatenated together. This is useful for performance benchmarking because the ideal final state is known (all qubits should return to their initial state), making deviations easy to detect.

```{code-cell} ipython3
qc_mirrored = get_benchmark(
    benchmark="dj",
    level=BenchmarkLevel.MAPPED,
    circuit_size=5,
    target=get_device("ionq_forte_36"),
    opt_level=2,
    # generate_mirror_circuit=True,
)
qc_mirrored.draw(output="mpl")
```

## Self-defined Circuits

> Sometimes you want to benchmark a circuit that you designed yourself. You can create a `qiskit.QuantumCircuit`, pass it directly to `get_benchmark`, and let `mqt.bench` compile it for your chosen target backend.

Reviewer: > 
```suggestion
Sometimes you want to derive benchmark circuits from an original circuit that you created yourself. You can create a `qiskit.QuantumCircuit`, pass it directly to `get_benchmark`, and let `mqt.bench` compile it for your chosen target backend.
```

```{code-cell} ipython3
from qiskit import QuantumCircuit

qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)

qc_circuit = get_benchmark(benchmark=qc, level=BenchmarkLevel.NATIVEGATES, target=get_target_for_gateset("ionq_forte", 2))
qc_circuit.draw(output="mpl")
```

## Self-defined Targets

You can also provide a target manually. Here we build a 5‑qubit `GenericBackendV2` with a custom basis and attach a short description before feeding it into `get_benchmark`.

```{code-cell} ipython3
from qiskit.providers.fake_provider import GenericBackendV2

standard_gates = ["id", "x", "sx", "rz", "cx"]
backend = GenericBackendV2(num_qubits=5, basis_gates=standard_gates)
target = backend.target
target.description = "Awesome Target"

qc_target = get_benchmark(
    benchmark="dj",
    level=BenchmarkLevel.NATIVEGATES,
    circuit_size=5,
    target=target,
    opt_level=2,
)
qc_target.draw(output="mpl")
```

## Random and Symbolic Parameters

> Benchmarks such as QAOA require sets of real parameters. `mqt.bench can either choose random numeric values for you (default) or leave the parameters symbolic so that you can optimize them later.

Reviewer: ```suggestion
Variational benchmarks such as QAOA require sets of real parameters. MQT Bench can either choose random numeric values for you (default) or leave the parameters symbolic so that you can optimize them later.
```

    level=BenchmarkLevel.MAPPED,
    circuit_size=5,
    target=get_device("ionq_forte_36"),
    opt_level=2,
)
qc_mapped_level.draw(output="mpl")
```

## Mirror Circuits

Mirror circuits replicate a given benchmark circuit and its mirror image (inverse) concatenated together. This is useful for performance benchmarking because the ideal final state is known (all qubits should return to their initial state), making deviations easy to detect. Here, we mirror the circuit obtained in the {ref}`mapped-level` section.

```{code-cell} ipython3
qc_mirrored = get_benchmark(
    benchmark="dj",
    level=BenchmarkLevel.MAPPED,
    circuit_size=5,
    target=get_device("ionq_forte_36"),
    opt_level=2,

>     # generate_mirror_circuit=True,

Reviewer 2: ```suggestion
    generate_mirror_circuit=True,
```
Just so that we do not forget.

target = backend.target
target.description = "Awesome Target"

qc_target = get_benchmark(
    benchmark="dj",
    level=BenchmarkLevel.NATIVEGATES,
    circuit_size=5,
    target=target,
    opt_level=2,
)
qc_target.draw(output="mpl")
```

## Random and Symbolic Parameters

Variational benchmarks such as QAOA require sets of real parameters. MQT Bench can either choose random numeric values for you (default) or leave the parameters symbolic so that you can optimize them later.

### Random Parameters (Default)

Random numeric values are generated by default.

```{code-cell} ipython3
qc_random = get_benchmark(benchmark="qaoa", level=BenchmarkLevel.ALG, circuit_size=2)
qc_random.draw(output="mpl")
```

### Symbolic Parameters

Set `random_parameters=False` to keep the parameters symbolic instead of sampling them.

```{code-cell} ipython3
qc_symbolic = get_benchmark(benchmark="qaoa", level=BenchmarkLevel.ALG, circuit_size=2, random_parameters=False)
qc_symbolic.draw(output="mpl")
```

## Output Formats

The `save_circuit` function lets you export circuits in several industry‑standard formats so that they can be run or analyzed by other tools.

> ### QASM2

Reviewer 2: ```suggestion
### OpenQASM2
```

```{code-cell} ipython3
from pathlib import Path
from mqt.bench.output import (
    OutputFormat,
    save_circuit,
)

qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)

save_circuit(qc, "qasm2", BenchmarkLevel.INDEP, output_format=OutputFormat.QASM2)
text_qasm2 = Path("qasm2.qasm").read_text()
print(text_qasm2)
Path("qasm2.qasm").unlink()
```

### QASM3

```{code-cell} ipython3
qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)

save_circuit(qc, "qasm3", BenchmarkLevel.INDEP, output_format=OutputFormat.QASM3)
text_qasm3 = Path("qasm3.qasm").read_text()
print(text_qasm3)
Path("qasm3.qasm").unlink()

> ```

Reviewer 2: Both cells here could benefit from custom output formatting.
See for example the QASM code cells in MQT Core https://mqt.readthedocs.io/projects/core/en/latest/mqt_core_ir.html#visualizing-circuits

This requires https://github.com/munich-quantum-toolkit/core/blob/bda3f78c39508315da6b0175a1f7d1b18949cb45/pyproject.toml#L315
as well as https://github.com/munich-quantum-toolkit/core/blob/bda3f78c39508315da6b0175a1f7d1b18949cb45/docs/mqt_core_ir.md?plain=1#L35-L38

Author: Thanks a lot for the suggestion, but is there anything else needed? 
The current output does not differ from the old one.

Author: In Chrome I also get the syntax-highlighted one, should have checked that before trying to debug it ^^

Reviewer 2: Could also be a caching issue 🧐
Anyway. I think this can be resolved 😌

### QPY

```{code-cell} ipython3
qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)

save_circuit(qc, "qpy", BenchmarkLevel.INDEP, output_format=OutputFormat.QPY)
text_qpy = Path("qpy.qpy").read_bytes()
print(text_qpy)
Path("qpy.qpy").unlink()

> ```

Reviewer 2: Printing a binary file is rather pointless, isn't it?
